ethcore/src/blockchain/blockchain.rs

```diff
@@ -979,6 +995,34 @@ impl BlockChain {
 		}
 	}
 
+	/// Mark a block to be considered finalized. Panic if the hash does not exist.
+	pub fn mark_finalized(&self, batch: &mut DBTransaction, block_hash: H256) {
+		let mut block_details = self.block_details(&block_hash).unwrap_or_else(|| panic!("Invalid block hash: {:?}", block_hash));
+		block_details.finalized = true;
+
+		self.prepare_details_update(batch, block_hash, block_details);
+	}
+
+	/// Update metadata detail for an existing block.
+	pub fn update_metadatas(&self, batch: &mut DBTransaction, block_hash: H256, metadatas: HashMap<Bytes, Bytes>) {
+		let mut block_details = self.block_details(&block_hash).unwrap_or_else(|| panic!("Invalid block hash: {:?}", block_hash));
+		for (key, value) in metadatas {
+			block_details.metadatas.insert(key, value);
+		}
+
+		self.prepare_details_update(batch, block_hash, block_details);
+	}
+
+	/// Prepares extras block detail update.
+	fn prepare_details_update(&self, batch: &mut DBTransaction, block_hash: H256, block_details: BlockDetails) {
```

Comment:
Does it just prepare the update or write it to cache? Seems there is already `prepare_block_details_update` method in this struct and it returns the `HashMap`, might be a bit misleading to have that one here as well.

Reply:
That was meant to be a stripped version of `prepare_update`, but yeah looks confusing. I'll change the name.